File: ViewBase/Source/Behaviors/DataGridBehavior.vb
Imports System.ComponentModel
Imports System.Windows
Imports System.Windows.Controls
Imports System.Windows.Data
Imports Wells.StandardModel.Attributes

''' <summary>
''' Using this behavior on a DataGrid will ensure to display only columns with "Browsable Attributes"
''' </summary>
Public Class DataGridBehavior

    Public Shared ReadOnly UseBrowsableAttributeOnColumnProperty As DependencyProperty =
            DependencyProperty.RegisterAttached("UseBrowsableAttributeOnColumn",
            GetType(Boolean),
            GetType(DataGridBehavior),
            New UIPropertyMetadata(False, AddressOf UseBrowsableAttributeOnColumnChanged))


    Public Shared Function GetUseBrowsableAttributeOnColumn(obj As DependencyObject) As Boolean
        Return CType(obj.GetValue(UseBrowsableAttributeOnColumnProperty), Boolean)
    End Function


    Public Shared Sub SetUseBrowsableAttributeOnColumn(obj As DependencyObject, val As Boolean)
        obj.SetValue(UseBrowsableAttributeOnColumnProperty, val)
    End Sub

    Private Shared Sub UseBrowsableAttributeOnColumnChanged(obj As DependencyObject, e As DependencyPropertyChangedEventArgs)
        Dim dataGrid = CType(obj, DataGrid)
        If dataGrid IsNot Nothing Then
            If CType(e.NewValue, Boolean) Then
                AddHandler dataGrid.AutoGeneratingColumn, AddressOf DataGridOnAutoGeneratingColumn
            Else
                RemoveHandler dataGrid.AutoGeneratingColumn, AddressOf DataGridOnAutoGeneratingColumn
            End If
        End If
    End Sub

    Private Shared Sub DataGridOnAutoGeneratingColumn(sender As Object, e As DataGridAutoGeneratingColumnEventArgs)
        Dim propDesc = CType(e.PropertyDescriptor, PropertyDescriptor)

        Dim changeStyle As Boolean = False
        If e.PropertyType Is GetType(Date) Then
            CType(e.Column, DataGridTextColumn).Binding.StringFormat = "dd/MM/yyyy"
            changeStyle = True
        ElseIf e.PropertyType Is GetType(Double) Then
            CType(e.Column, DataGridTextColumn).Binding.StringFormat = "N3"
            changeStyle = True
        ElseIf e.PropertyType Is GetType(Integer) Then
            CType(e.Column, DataGridTextColumn).Binding.StringFormat = "N0"
            changeStyle = True
        End If

        If changeStyle Then
            Dim style = New Style(GetType(TextBlock), CType(e.Column, DataGridTextColumn).ElementStyle)
            Dim setter = New Setter(FrameworkElement.HorizontalAlignmentProperty, HorizontalAlignment.Right)
            style.Setters.Add(setter)
            CType(e.Column, DataGridTextColumn).ElementStyle = style
        End If

        If propDesc IsNot Nothing Then
            For Each att In propDesc.Attributes
                If TypeOf att Is BrowsableAttribute Then
                    Dim browsableAttribute = CType(att, BrowsableAttribute)

                    If browsableAttribute IsNot Nothing Then
                        Dim isBrowsable = browsableAttribute.Browsable
                        If Not isBrowsable Then
                            e.Cancel = Not isBrowsable
                            Exit Sub
                        End If
                    End If
                End If

                If TypeOf att Is DisplayNameAttribute Then
                    Dim displayName = CType(att, DisplayNameAttribute)
                    If displayName IsNot Nothing Then
                        e.Column.Header = displayName.DisplayName
                    End If
                End If

                If TypeOf att Is SortIndexAttribute Then
                    Dim sortIndex = CType(att, SortIndexAttribute)
                    If sortIndex IsNot Nothing Then
                        e.Column.DisplayIndex = sortIndex.Index
                    End If
                End If
            Next
        End If

        If e.PropertyType.IsEnum Then
            CType(e.Column, DataGridComboBoxColumn).ItemsSource = Base.Common.EnumDescriptionsToList(e.PropertyType)
            Dim propPath = CType(e.Column, DataGridComboBoxColumn).SortMemberPath
            CType(e.Column, DataGridComboBoxColumn).SelectedItemBinding = New Binding With {
                    .Path = New PropertyPath(propPath),
                    .Converter = New EnumValueConverter,
                    .Mode = BindingMode.TwoWay,
                    .UpdateSourceTrigger = UpdateSourceTrigger.PropertyChanged
                }
        End If
    End Sub

End Class

File: BaseView/src/Behaviors/DataGridBehavior.cs
﻿using System;
using System.ComponentModel;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Data;
using Wells.Base;
using Wells.BaseModel.Attributes;
using Wells.BaseView.Converters;

namespace Wells.BaseView.Behaviors
{
    public class DataGridBehavior
    {

        public static readonly DependencyProperty UseBrowsableAttributeOnColumnProperty =
             DependencyProperty.RegisterAttached("UseBrowsableAttributeOnColumn",
            typeof(bool),
            typeof(DataGridBehavior),
            new UIPropertyMetadata(false, UseBrowsableAttributeOnColumnChanged));

        public static bool GetUseBrowsableAttributeOnColumn(DependencyObject obj)
        {
            return (bool)obj.GetValue(UseBrowsableAttributeOnColumnProperty);
        }

        public static void SetUseBrowsableAttributeOnColumn(DependencyObject obj, bool val)
        {
            obj.SetValue(UseBrowsableAttributeOnColumnProperty, val);
        }


        private static void UseBrowsableAttributeOnColumnChanged(DependencyObject obj, DependencyPropertyChangedEventArgs e)
        {
            var dataGrid = (DataGrid)obj;
            if (dataGrid != null)
            {
                if ((bool)e.NewValue)
                {
                    dataGrid.AutoGeneratingColumn += DataGridOnAutoGeneratingColumn;
                }
                else
                {
                    dataGrid.AutoGeneratingColumn -= DataGridOnAutoGeneratingColumn;
                }
            }
        }


        private static void DataGridOnAutoGeneratingColumn(object sender, DataGridAutoGeneratingColumnEventArgs e)
        {
            var propDesc = (PropertyDescriptor)e.PropertyDescriptor;

            bool changeStyle = false;
            if (e.PropertyType == typeof(DateTime))
            {
                (e.Column as DataGridTextColumn).Binding.StringFormat = "dd/MM/yy";
                changeStyle = true;
            }
            if (Common.IsNumericType(e.PropertyType))
            {
                if (Common.IsIntegerNumericType(e.PropertyType))
                {
                    (e.Column as DataGridTextColumn).Binding.StringFormat = "N0";
                }
                else
                {
                    (e.Column as DataGridTextColumn).Binding.StringFormat = "N3";
                }
                changeStyle = true;
            }
         
            if (changeStyle)
            {
                ((e.Column as DataGridTextColumn).Binding as Binding).ConverterCulture = System.Globalization.CultureInfo.CurrentCulture;
                var style = new Style(typeof(TextBlock), (e.Column as DataGridTextColumn).ElementStyle);
                var setter = new Setter(FrameworkElement.HorizontalAlignmentProperty, HorizontalAlignment.Right);
                style.Setters.Add(setter);
                (e.Column as DataGridTextColumn).ElementStyle = style;
            }

            bool thereAreColumns = (sender as DataGrid).Columns.Count > 0;

            if (propDesc != null)
            {
                foreach (var att in propDesc.Attributes)
                {
                    if (att is BrowsableAttribute browsableAttribute && !browsableAttribute.Browsable)
                    {
                        e.Cancel = !browsableAttribute.Browsable;
                        return;
                    }

                    if (att is DisplayNameAttribute displayName)
                    {
                        e.Column.Header = displayName.DisplayName;
                    }

                    if (thereAreColumns && att is SortIndexAttribute sortIndex)
                    {
                        e.Column.DisplayIndex = (int)sortIndex.Index;
                    }
                }
            }


            if (e.PropertyType.IsEnum)
            {
                (e.Column as DataGridComboBoxColumn).ItemsSource = Base.Common.EnumDescriptionsToList(e.PropertyType);
                var propPath = (e.Column as DataGridComboBoxColumn).SortMemberPath;
                (e.Column as DataGridComboBoxColumn).SelectedItemBinding = new Binding
                {
                    Path = new PropertyPath(propPath),
                    Converter = new EnumValueConverter(),
                    Mode = BindingMode.TwoWay,
                    UpdateSourceTrigger = UpdateSourceTrigger.PropertyChanged
                };
            }

        }
    }
}
